Improve shrinking of code using flatmap/composite

DRMacIver opened the pull request:
In existing Hypothesis there is a problem where often code
using these results in examples that don't shrink very well
because shrinking an earlier value results in changes to
the structure of the later value that no longer exhibits
the problematic behaviour.

This change introduces some limited ability to work past
that by allowing the later structure to be reordered in a
way that gives more flexible possibilities for producing
shrunk examples..

This is primarily focused on cases that use length prefixed
lists and similar concepts (hence the tests focusing on that),
but likely should work more generally.

This branch also contains a number of other improvements to
the Conjecture engine that turned out to be necessary to make
this work well, and one removal of an annoying shrinking pass
that rarely worked in the first place and never worked in the
presence of some of the other improvements that I ended up
making to it.

DRMacIver commented:
PS. @alexwlchan, you'll probably want to read [How Hypothesis Works](http://hypothesis.works/articles/how-hypothesis-works/) before reviewing this.

alexwlchan commented on src/hypothesis/internal/conjecture/data.py:
```diff
@@ -105,6 +106,19 @@ def draw(self, strategy):
+        as we run a pass which tries to shrink the byte stream prior to a bind
+        point while rearranging what comes after somewhat to allow for more
+        flexibility. Trying that at every point in the data stream is
+        prohibitvely expensive, but trying it at a couple dozen is basically
```
/s/prohibitvely/prohibitively

DRMacIver commented on src/hypothesis/internal/conjecture/data.py:
```diff
@@ -105,6 +106,19 @@ def draw(self, strategy):
+        flexibility. Trying that at every point in the data stream is
+        prohibitively expensive, but trying it at a couple dozen is basically
+        fine."""
+        self.bind_points.add(self.index)
```
Correct.

DRMacIver commented on src/hypothesis/internal/conjecture/engine.py:
```diff
@@ -465,24 +480,63 @@ def replace(b):
+                interval_counter = self.changed
+                i = 0
+                alternatives = None
+                while i < len(self.last_data.intervals):
```
There's actually an important difference, and it's that "ubiquitous use of side effects" thing again (it really is a good way to write this once you get used to the style, I promise. It's actually quite robust and the alternatives are all much more unwieldy).

last_data can change in the course of running this. When it changes, this may cause the number of intervals to change. So using hrange() here would be incorrect because you can suddenly find yourself with an index out of the range of the current set of intervals.

It is a deliberate feature that sometimes the set of intervals prior to i may have changed and we might skip them over (that's one of the reasons why running this to fixation helps) but that we still stay at the old index - it's a good heuristic for avoiding performing O(n^2) useless shrinks that would happen if we restarted the loop every time.

DRMacIver commented on src/hypothesis/internal/conjecture/minimizer.py:
```diff
@@ -82,15 +83,38 @@ def run(self):
+                    break
+
+            for c in sorted(set(self.current), reverse=True):
+                for d in range(c):
```
Yes, thanks.

DRMacIver commented on src/hypothesis/strategies.py:
```diff
@@ -901,7 +901,15 @@ def accept(*args, **kwargs):
+                def draw(strategy):
+                    if not first_draw[0]:
+                        data.mark_bind()
+                    first_draw[0] = False
```
Because it gets updated inside the draw() function, and because Python 2 compatibility we can't declare it nonlocal.

DRMacIver commented:
> One thing that wasn’t entirely obvious: what’s the purpose of the cautious flag? It looks like it causes us to skip trying the small shrinks – are they just not appropriate and/or slow in these contexts?

No, there's an actual reason they'd be bad to do - we still try small shrinks which would cause any single value to be reduced (some of the new code in the minimizer does that now - the bit that does a delta debugging style pass for lowering bytes), but we skip the ones that can ever cause a byte value to *increase*.

The theory is that Hypothesis is trying to produce a lexicographically minimal buffer amongst the buffers of shortest size. However the practice is that there are some shapes of data where this will mess you up. [This test](https://github.com/HypothesisWorks/hypothesis-python/blob/master/tests/nocover/test_example_quality.py#L567) is an example that can trigger it.

The problem case is basically that some times you can get lexicographic shrinks that produce "worse" examples because they shrunk a bit that turned out not to matter very much but made a bit that mattered much more very complex. Those shrinks will almost always then shrink further into something nice, but you end up with this annoying cycle where the data goes simple -> complex -> simple -> complex, and if you're unlucky Hypothesis stops in the middle of a complex cycle and you get a bad shrink.

These come because the value `[0, 255]` is lexicographically simpler than  `[1, 0]`. But if we've got something like drawing from `booleans()` then `integers()` then `[1, 0, 0, 0, 0, 0, 0, 0, 0]` gives `(True, 0)` while, `[0, 255, 0, 0, 0, 0, 0, 0, 0]` gives `(False, 18374686479671623680)`. So you get things that are simpler by Hypothesis's internal heuristics but more complex to a user, and you don't really want to stop at that point.

OTOH there are cases where these shrinks are absolutely necessary: For example, they're what enable Hypothesis to shrink the integer 'n' to 'n - 1'.

The way this has historically been handled is that the lexicographic minimizer was only run within atomic blocks in the data stream (things that are the result of a single call to draw_bytes) - this ended up being a nice halfway house where it mostly shrunk the things it needed to shrink, but didn't ever cause these pathological shrinks.

The reason this changes in this pull request is that I wanted the lexicographic minimizer to be able to run on whole prefixes: The matrices test is an example where you need to be able to perform lexicographic minimization on a whole prefix instead of just blocks (they're a pair of adjacent blocks in this case, but that doesn't actually help - a pair of adjacent blocks is enough to trigger the problem behaviour).

The reason that this is then used more widely is that only minimizing blocks is kinda rubbish and loses out on a bunch of things that can help us out - both improving example quality and getting faster shrinks

alexwlchan commented on src/hypothesis/internal/conjecture/minimizer.py:
```diff
@@ -151,6 +151,17 @@ def run(self):
 
 def minimize(initial, condition, random=None, cautious=False):
+    """Perform a lexicographical minimization of the byte string 'initial' such
+    that the predicate 'condition' returns True, and return the minmized
```
/s/minmized/minimized

DRMacIver commented:
I've fixed all the uses of bare range (we should see if we can lint that somehow), plus a few more besides in conjecture. Have also rebased to squash into a tidier set of commits.

DRMacIver commented:
Whoops, I fucked *that* up. Now pushed again with less completely failing to import the thing I was using.

DRMacIver commented:
BTW @alexwlchan I think this now has all changes you've requested but still requires you to rubber stamp the review.

alexwlchan commented:
Yep, I’ve been re-reading it this morning.

I’ll give it the green tick, but as you’re about to do a new release, I’ll let you decide if you want to merge it before or after. (I can’t think why you wouldn’t want to release immediately, but might as well ask.)

DRMacIver commented:
Yeah, I tend to release things immediately once they're ready unless either there's a very good reason not to or I forget (as per the fake factory thing).

However I think there's another change I'll want to get in before the release, so I may have another PR for you to look at before I release.

alexwlchan commented:
Fine, merged. (I’ve also got a bit more progress on the `one_of` stuff, so I’ll see if I can’t get that finished this afternoon.)
